pkg/framework/machinesets.go

```diff
@@ -369,6 +449,106 @@ func WaitForMachineSetDelete(c runtimeclient.Client, machineSet *machinev1.Machi
+func WaitForSpotMachineSet(c client.Client, name string) error {
+	machineSet, err := GetMachineSet(c, name)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Retry until the MachineSet is ready.
+	err = wait.PollImmediate(RetryMedium, WaitLong, func() (bool, error) {
+		machines, err := GetMachinesFromMachineSet(c, machineSet)
+		if err != nil {
+			return true, err
+		}
+
+		replicas := pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0)
+		if len(machines) != int(replicas) {
+			klog.Infof("%q: found %d Machines, but MachineSet has %d replicas", name, len(machines), int(replicas))
+			return false, nil
+		}
+
+		failed := FilterMachines(machines, MachinePhaseFailed)
+		if len(failed) > 0 {
+			// if there are failed machines, print them out before we exit
+			klog.Errorf("found %d Machines in failed phase: ", len(failed))
+			for _, m := range failed {
+				reason := "failureReason not present in Machine.status"
+				if m.Status.ErrorReason != nil {
+					reason = string(*m.Status.ErrorReason)
+				}
+				message := "failureMessage not present in Machine.status"
+				if m.Status.ErrorMessage != nil {
+					message = string(*m.Status.ErrorMessage)
+				}
+				klog.Errorf("Failed machine: %s, Reason: %s, Message: %s", m.Name, reason, message)
+			}
+		}
+		Expect(len(failed)).To(Equal(0))
+
+		// Check if any machine did not get provisioned because of insufficient spot capacity.
+		for _, m := range machines {
+			if hasInsufficientCapacity(m, platform) {
+				return true, ErrMachineNotProvisionedInsufficientCloudCapacity
+			}
+		}
+
+		running := FilterRunningMachines(machines)
+		// This could probably be smarter, but seems fine for now.
+		if len(running) != len(machines) {
+			klog.Infof("%q: not all Machines are running: %d of %d", name, len(running), len(machines))
+			return false, nil
+		}
+
+		for _, m := range running {
+			node, err := GetNodeForMachine(c, m)
+			if err != nil {
+				klog.Infof("Node for machine %s not found yet: %v", m.Name, err)
+				return false, nil
+			}
+
+			if !IsNodeReady(node) {
+				klog.Infof("%s: node is not ready", node.Name)
+				return false, nil
+			}
+		}
+
+		return true, nil
+	})
+	return err
+}
+
+// hasInsufficientCapacity return true if the machine cannot be provisioned due to insufficient spot capacity.
+func hasInsufficientCapacity(m *machinev1.Machine, platform configv1.PlatformType) bool {
+	switch platform {
+	case configv1.AWSPlatformType:
+		awsProviderStatus := machinev1.AWSMachineProviderStatus{}
+		if m.Status.ProviderStatus != nil {
+			err := json.Unmarshal(m.Status.ProviderStatus.Raw, &awsProviderStatus)
+			Expect(err).ToNot(HaveOccurred())
+			for _, condition := range awsProviderStatus.Conditions {
+				if condition.Type == string(machinev1.MachineCreation) && condition.Status == metav1.ConditionFalse && strings.Contains(condition.Message, "InsufficientInstanceCapacity") {
```

Comment:
nit: I think about moving error texts to constants, wdyt?

Reply:
Maybe generalize this condition and extract it to some function, instead of having two quite similar but slightly different looong lines?